### PR TITLE
Invariant capable records

### PR DESCRIPTION
**Pull request by RichardIrons-neo4j**
This changes the Record type so that it does not contain a dictionary on every instance, instead only  containing the array of values, and references to two dictionaries shared across the whole cursor: one just a straight Dictionary<string, object>, and the other of the same type but using an InvariantCultureCaseInsensitive string comparer, so that values can be looked up using a case insensitive key, used in the object mapping. 

Added Get<T> and TryGet<T> methods to IRecord and IEntity, and added xxxCaseInsensitive versions of those methods to IRecord.

More tests added, tests modified to create records through `TestRecord.Create`.

### CODE REVIEW

File: Neo4j.Driver/Neo4j.Driver/Internal/Result/ResultCursorBuilder.cs [hunk 1]
```diff
@@ -139,7 +144,11 @@ public async ValueTask<IResultSummary> ConsumeAsync()
     public void RunCompleted(long queryId, string[] fields, IResponsePipelineError error)
     {
         _queryId = queryId;
-        _fields = fields;
+
+        _fieldLookup = fields.Select((field, index) => (field, index))
```
Comment (thelonelyvulpes) [inline Neo4j.Driver/Neo4j.Driver/Internal/Result/ResultCursorBuilder.cs]: can we do this the old fashioned way, and do it in 1 pass.
```
var invariant = new Dictionary<string, int>(fields.Count, StringComparer.InvariantCultureIgnoreCase);
var exact = new Dictionary<string, int>(fields.Count);
for (var i = 0; i < fields.Count; i++)
{
    invariant.Add(fields[i], i);
    exact.Add(fields[i], i);
}
_invariantFieldLookup = invariant;
_fieldLookup = exact;
```

Comment (RichardIrons-neo4j) [reply Neo4j.Driver/Neo4j.Driver/Internal/Result/ResultCursorBuilder.cs]: Done

File: Neo4j.Driver/Neo4j.Driver/Internal/Result/Record.cs [hunk 2]
```diff
@@ -13,34 +13,118 @@
...
+    /// <inheritdoc />
+    IEnumerable<string> IReadOnlyDictionary<string, object>.Keys => Keys;
+
+    /// <inheritdoc />
+    IEnumerable<object> IReadOnlyDictionary<string, object>.Values => _fieldValues;
+
+    /// <inheritdoc />
+    IEnumerator<KeyValuePair<string, object>> IEnumerable<KeyValuePair<string, object>>.GetEnumerator()
+    {
+        return Keys.Select(key => new KeyValuePair<string, object>(key, this[key])).GetEnumerator();
```
Comment (thelonelyvulpes) [inline Neo4j.Driver/Neo4j.Driver/Internal/Result/Record.cs]:
```suggestion
        return Keys.Select((key, i) => new KeyValuePair<string, object>(key, _fieldValues[i])).GetEnumerator();
```

File: Neo4j.Driver/Neo4j.Driver/Internal/Result/ResultCursorBuilder.cs [hunk 3]
```diff
@@ -139,7 +144,20 @@ public async ValueTask<IResultSummary> ConsumeAsync()
     public void RunCompleted(long queryId, string[] fields, IResponsePipelineError error)
     {
         _queryId = queryId;
-        _fields = fields;
+
+        if (fields is not null)
```
Comment (thelonelyvulpes) [inline Neo4j.Driver/Neo4j.Driver/Internal/Result/ResultCursorBuilder.cs:148]: We probably don't need this check. If we want to include it, we should instead throw a invalid or protocol exception if fields is null

File: Neo4j.Driver/Neo4j.Driver/Public/IRecord.cs [hunk 4]
```diff
@@ -28,11 +28,49 @@ public interface IRecord
     /// <summary>Gets the value specified by the given key.</summary>
     /// <param name="key">The key</param>
     /// <returns>the value specified with the given key.</returns>
-    object this[string key] { get; }
+    new object this[string key] { get; }
```
Comment (thelonelyvulpes) [inline Neo4j.Driver/Neo4j.Driver/Public/IRecord.cs]: we could drop this as it is defined on the IReadOnlyDictionary<string, object> interface and behaves the same.

File: Neo4j.Driver/Neo4j.Driver/Internal/Result/Record.cs (hunk 2 above)
Comment (RichardIrons-neo4j) [reply Neo4j.Driver/Neo4j.Driver/Internal/Result/Record.cs]: Done.

File: Neo4j.Driver/Neo4j.Driver/Public/IRecord.cs (hunk 4 above)
Comment (RichardIrons-neo4j) [reply Neo4j.Driver/Neo4j.Driver/Public/IRecord.cs]: Done.

File: Neo4j.Driver/Neo4j.Driver/Internal/Result/ResultCursorBuilder.cs (hunk 3 above)
Comment (RichardIrons-neo4j) [reply Neo4j.Driver/Neo4j.Driver/Internal/Result/ResultCursorBuilder.cs:148]: There are a couple of places where null is passed in as the `fields` parameter - `OnFailure` and `OnIgnored`. The old code didn't use to do anything with the array so the check wasn't needed, but now it builds the lookups it needs the check. We could change those places to pass an empty array but then we'd have it creating two empty dictionaries for no reason.

File: Neo4j.Driver/Neo4j.Driver/Preview/Mapping/DictAsRecord.cs [hunk 5]
```diff
@@ -27,20 +28,91 @@ internal class DictAsRecord : IRecord
 
     public DictAsRecord(object dict, IRecord record)
```
Comment (thelonelyvulpes) [inline Neo4j.Driver/Neo4j.Driver/Preview/Mapping/DictAsRecord.cs:29]: can we make this class sealed

File: Neo4j.Driver/Neo4j.Driver/Internal/Result/Record.cs [hunk 6]
```diff
@@ -13,34 +13,100 @@
...
 // limitations under the License.
 
+using System.Collections;
 using System.Collections.Generic;
+using System.Linq;
 
 namespace Neo4j.Driver.Internal.Result;
 
 internal class Record : IRecord
 {
```
Comment (thelonelyvulpes) [inline Neo4j.Driver/Neo4j.Driver/Internal/Result/Record.cs:23]: let's make this class sealed

File: Neo4j.Driver/Neo4j.Driver/Preview/Mapping/DictAsRecord.cs [hunk 7]
```diff
@@ -27,20 +28,91 @@ internal class DictAsRecord : IRecord
...
+            IEntity entity => entity.Properties ?? new Dictionary<string, object>(),
             IReadOnlyDictionary<string, object> dictionary => dictionary,
             _ => throw new InvalidOperationException($"Cannot create a DictAsRecord from a {dict.GetType().Name}")
         };
 
+        // this is only used by the default mapper so make it case insensitive
+        _dict = readOnlyDictionary.ToDictionary(
+            kvp => kvp.Key,
+            kvp => kvp.Value,
+            StringComparer.OrdinalIgnoreCase);
```
Comment (thelonelyvulpes) [inline Neo4j.Driver/Neo4j.Driver/Preview/Mapping/DictAsRecord.cs]: This is ordinal but the other is CultureInvariantIgnoreCase, we should only use one for consistency.

File: Neo4j.Driver/Neo4j.Driver/Preview/Mapping/DictAsRecord.cs (hunk 5 above)
Comment (RichardIrons-neo4j) [reply Neo4j.Driver/Neo4j.Driver/Preview/Mapping/DictAsRecord.cs:29]: done

File: Neo4j.Driver/Neo4j.Driver/Internal/Result/Record.cs (hunk 6 above)
Comment (RichardIrons-neo4j) [reply Neo4j.Driver/Neo4j.Driver/Internal/Result/Record.cs:23]: done

File: Neo4j.Driver/Neo4j.Driver/Preview/Mapping/DictAsRecord.cs (hunk 7 above)
Comment (RichardIrons-neo4j) [reply Neo4j.Driver/Neo4j.Driver/Preview/Mapping/DictAsRecord.cs]: Changed to CultureInvariantIgnoreCase

**Review by thelonelyvulpes — APPROVED** [review]

**Pull request merged**